src/lib/tls/tls_algos.h

```diff
@@ -117,48 +118,81 @@ enum class Group_Params : uint16_t {
+
+      constexpr bool operator<(Group_Params other) const { return m_code < other.m_code; }
+
+      constexpr Group_Params_Code code() const { return m_code; }
+
+      constexpr uint16_t wire_code() const { return static_cast<uint16_t>(m_code); }
+
+      constexpr bool is_x25519() const { return m_code == Group_Params_Code::X25519; }
 
-std::string group_param_to_string(Group_Params group);
-Group_Params group_param_from_string(std::string_view group_name);
-std::vector<std::pair<std::string, std::string>> hybrid_group_param_to_algorithm_specs(Group_Params group);
-bool group_param_is_dh(Group_Params group);
+      constexpr bool is_ecdh_named_curve() const {
+         return m_code == Group_Params_Code::SECP256R1 || m_code == Group_Params_Code::SECP384R1 ||
+                m_code == Group_Params_Code::SECP521R1 || m_code == Group_Params_Code::BRAINPOOL256R1 ||
+                m_code == Group_Params_Code::BRAINPOOL384R1 || m_code == Group_Params_Code::BRAINPOOL512R1;
+      }
+
+      constexpr bool is_dh_named_group() const {
+         return m_code == Group_Params_Code::FFDHE_2048 || m_code == Group_Params_Code::FFDHE_3072 ||
+                m_code == Group_Params_Code::FFDHE_4096 || m_code == Group_Params_Code::FFDHE_6144 ||
+                m_code == Group_Params_Code::FFDHE_8192;
+      }
+
+      constexpr bool is_diffie_hellman() const {
+         const uint16_t group_id = wire_code();
+         return (group_id >= 256 && group_id < 512);
+      }
+
+      constexpr bool is_pure_kyber() const {
+         return m_code == Group_Params_Code::KYBER_512_R3 || m_code == Group_Params_Code::KYBER_768_R3 ||
+                m_code == Group_Params_Code::KYBER_1024_R3;
+      }
+
+      constexpr bool is_post_quantum() const { return is_pure_kyber() || is_pqc_hybrid(); }
+
+      constexpr bool is_pqc_hybrid() const {
+         return m_code == Group_Params_Code::HYBRID_X25519_KYBER_512_R3_CLOUDFLARE ||
+                m_code == Group_Params_Code::HYBRID_X25519_KYBER_768_R3_CLOUDFLARE ||
+                m_code == Group_Params_Code::HYBRID_X25519_KYBER_512_R3_OQS ||
+                m_code == Group_Params_Code::HYBRID_X25519_KYBER_768_R3_OQS ||
+                m_code == Group_Params_Code::HYBRID_SECP256R1_KYBER_512_R3_OQS ||
+                m_code == Group_Params_Code::HYBRID_SECP384R1_KYBER_768_R3_OQS ||
+                m_code == Group_Params_Code::HYBRID_SECP521R1_KYBER_1024_R3_OQS;
+      }
+
+      constexpr bool is_kem() const { return is_pure_kyber() || is_pqc_hybrid(); }
+
+      // Returns std::nullopt if the param has no known name
+      std::optional<std::string> to_string() const;
+
+      Group_Params(const Group_Params& other) = default;
+      Group_Params(Group_Params&& other) = default;
+      Group_Params& operator=(const Group_Params&) = default;
+      Group_Params& operator=(Group_Params&&) = default;
+      ~Group_Params() = default;
```

Comment:
If all are default, none needs to be defined. No hard feelings either way.

Reply:
:+1: removed